pkg/runtime/init.go

```diff
@@ -13,3 +13,263 @@
+	RemoteApplyYaml                = `echo '%s' | kubectl apply -f -`
+	RemoteCmdGetNetworkInterface   = "ls /sys/class/net"
+	RemoteCmdExistNetworkInterface = "ip addr show %s | egrep \"%s\" || true"
+	WriteKubeadmConfigCmd          = "cd %s && echo \"%s\" > kubeadm-config.yaml"
+	DefaultVIP                     = "10.103.97.2"
+	DefaultAPIserverDomain         = "apiserver.cluster.local"
+	DefaultRegistryPort            = 5000
+)
+
+func (k *KubeadmRuntime) loadMetadata() error {
+	metadata, err := LoadMetadata(filepath.Join(k.getCloudImageDir(), common.DefaultMetadataName))
+	if err != nil {
+		return err
+	}
+	k.Metadata = metadata
+	return nil
+}
+
+func (k *KubeadmRuntime) ConfigKubeadmOnMaster0() error {
+	if err := k.LoadFromClusterfile(k.RuntimeConfig.Clusterfile); err != nil {
+		return fmt.Errorf("failed to load kubeadm config from clusterfile: %v", err)
+	}
+	// TODO handle the kubeadm config, like kubeproxy config
+
+	bs, err := k.Merge(k.getDefaultKubeadmConfig())
+	if err != nil {
+		return fmt.Errorf("failed to merge kubeadm config: %v", err)
+	}
+
+	if err := utils.WriteFile(k.getDefaultKubeadmConfig(), bs); err != nil {
+		return fmt.Errorf("failed to dump new kubeadm config: %v", err)
+	}
+	return nil
+}
+
+func (k *KubeadmRuntime) getCertSANS() []string {
+	return k.ClusterConfiguration.APIServer.CertSANs
+}
+
+//CmdToString is in host exec cmd and replace to spilt str
+func (k *KubeadmRuntime) CmdToString(host, cmd, split string) string {
+	ssh, err := k.getHostSSHClient(host)
+	if err != nil {
+		logger.Error("failed to get host ssh client, %s %v", cmd, err)
+		return ""
+	}
+	data, err := ssh.Cmd(host, cmd)
+	if err != nil {
+		logger.Error("exec remote cmd failed, %s %v", cmd, err)
+	}
+	if data != nil {
+		str := string(data)
+		str = strings.ReplaceAll(str, "\r\n", split)
+		return str
+	}
+	return ""
+}
+
+func (k *KubeadmRuntime) getRemoteHostName(hostIP string) string {
+	hostName := k.CmdToString(hostIP, "hostname", "")
+	return strings.ToLower(hostName)
+}
+
+func (k *KubeadmRuntime) GenerateCert() error {
+	err := cert.GenerateCert(
+		k.getCertPath(),
+		k.getEtcdCertPath(),
+		k.getCertSANS(),
+		k.getMaster0IP(),
+		k.getRemoteHostName(k.getMaster0IP()),
+		k.getSvcCIDR(),
+		k.getDNSDomain(),
+	)
+	if err != nil {
+		return fmt.Errorf("generate certs failed %v", err)
+	}
+	k.sendNewCertAndKey([]string{k.getMaster0IP()})
+
+	return nil
+}
+
+func (k *KubeadmRuntime) CreateKubeConfig() error {
+	hostname := k.getRemoteHostName(k.getMaster0IP())
+	certConfig := cert.Config{
+		Path:     k.getCertPath(),
+		BaseName: "ca",
+	}
+
+	controlPlaneEndpoint := fmt.Sprintf("https://%s:6443", k.getAPIServerDomain())
+	err := cert.CreateJoinControlPlaneKubeConfigFiles(k.getBasePath(),
+		certConfig, hostname, controlPlaneEndpoint, "kubernetes")
+	if err != nil {
+		return fmt.Errorf("generator kubeconfig failed %s", err)
+	}
+	return nil
+}
+
+func (k *KubeadmRuntime) CopyStaticFiles(nodes []string) error {
+	var flag bool
```

Comment:
I am wondering if we need to add a lock to protect the race section since several goroutine would read/write the global `flag`.

Reply:
I will put all the goroutine error into channel.